apm-lambda-extension/extension/apm_server.go

```diff
@@ -48,30 +44,70 @@ const (
 
 // A struct to track the state and status of sending
 // to the APM server. Used in the backoff implementation.
-type ApmServerTransportStateType struct {
+type ApmServerTransport struct {
 	sync.Mutex
+	bufferPool        sync.Pool
+	ctx               context.Context
+	config            *extensionConfig
+	AgentDoneSignal   chan struct{}
+	DataChannel       chan AgentData
+	Client            *http.Client
 	Status            ApmServerTransportStatusType
 	ReconnectionCount int
 	GracePeriodTimer  *time.Timer
 }
 
-// The status of transport to the APM server.
-//
-// This instance of the ApmServerTransportStateType is public for use in tests.
-var ApmServerTransportState = ApmServerTransportStateType{
-	Status:            Healthy,
-	ReconnectionCount: -1,
+func InitApmServerTransport(ctx context.Context, config *extensionConfig) *ApmServerTransport {
+	var transport ApmServerTransport
+	transport.bufferPool = sync.Pool{New: func() interface{} {
+		return &bytes.Buffer{}
+	}}
+	transport.AgentDoneSignal = make(chan struct{}, 1)
+	transport.DataChannel = make(chan AgentData, 100)
+	transport.Client = &http.Client{
+		Timeout:   time.Duration(config.DataForwarderTimeoutSeconds) * time.Second,
+		Transport: http.DefaultTransport.(*http.Transport).Clone(),
+	}
+	transport.config = config
+	transport.ctx = ctx
+	transport.Status = Healthy
+	transport.ReconnectionCount = -1
+	return &transport
+}
+
+// StartBackgroundApmDataForwarding Receive agent data as it comes in and post it to the APM server.
+// Stop checking for, and sending agent data when the function invocation
+// has completed, signaled via a channel.
+func StartBackgroundApmDataForwarding(transport *ApmServerTransport, funcDone chan struct{}, backgroundDataSendWg *sync.WaitGroup) {
```

Comment:
Not a go expert but maybe this and other methods in this class should be receiver functions

```suggestion
func(transport *ApmServerTransport) StartBackgroundApmDataForwarding(funcDone chan struct{}, backgroundDataSendWg *sync.WaitGroup) {
```

Reply:
Implemented in 0855915.